### PR TITLE
Add Cognito User Pool Triggers

### LINKED ISSUE #3637: AWS Cognito User Pool triggers support
# This is a Feature Proposal

## Description
I'd like to be able to create functions that hook into a User Pool for various use-cases such as `preSignup`, `customMessage` etc. (which are defined [here](http://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html)). 

Currently I deploy the functions then manually attach them via AWS Console (although this can probably be automated using the AWS SDK).

I am aware that Lambda event sources are different from Lambda triggers (at least conceptually), but in practice they both seem to be only fired when an event occurs. Therefore, it would be lovely if I could define them in `serverless.yml` with something as follows:

```yml
functions:
  preSignup:
    handler: handler.preSignup
    events:
      - cognito-user-pool:
          pool: <POOL-ID>
          trigger: PreSignUp
```

Similar or dependent issues:
* #3400

### PR DESCRIPTION
**Pull request by hassankhan**
## What did you implement:

Closes #3637. Adds support for Cognito User Pool Triggers, and includes integration tests.

## How did you implement it:

Created new plugin for the `package:compileEvents` hook that checks `events` in `serverless.yml` for a ` cognitoUserPool` property.

## How can we verify it:

By running the integration tests in `tests/integration/aws/cognito-user-pool-trigger/`:

```bash
$ node_modules/.bin/jest cognito-user-pool-trigger
```

## Todos:

- [X] Write tests
- [X] Write documentation
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by HeroesDieYoung** [comment]
Any word on the priority of this?  I'm looking to have lambda's using cognito triggers in my project and would very much like to be able to configure that directly in serverless rather than having to go to the console for it.

**Comment by pmuens** [comment]
> Any word on the priority of this?

We've added it to the `v1.15` milestone (today we're releasing `v1.14`). We'll review this as soon as `v1.14` is out the door so that it can be merged ASAP.

Once merged you might want to work with `master` / the commit hash after merging so that you can already benefit w/o having to wait for `v1.15`.

**Comment by HeroesDieYoung** [comment]
Fantastic, thanks for the prompt reply!  I'll keep an eye on this PR then.

**Comment by hassankhan** [comment]
The milestone says `v1.15` should be done by June 7th, so I'd check back then @HeroesDieYoung 😉

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      const userPoolTemplate = {
+        Type: 'AWS::Cognito::UserPool',
+        Properties: {
+          UserPoolName: poolName,
+          LambdaConfig: lambdaConfig,
+        },
+      };
+
+      // Check if pool name has been declared in `Resources`
+      if (_.has(
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: Should we maybe check for the `Type` as well? Might be pretty unlikely but I'm thinking about the case when one has defined another Resource (e.g. an S3 bucket) with the name of a pool.

File: lib/plugins/create/templates/aws-csharp/serverless.yml [hunk 2]
```diff
@@ -79,6 +79,9 @@ functions:
 #              state:
 #                - pending
 #      - cloudwatchLog: '/aws/lambda/hello'
+#      - cognitoUserPool:
```
Comment (pmuens) [inline lib/plugins/create/templates/aws-csharp/serverless.yml]: An example should be added to the `serverless.yml.md` file as well

File: docs/providers/aws/events/cognito-user-pool-trigger.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
+<!--
+title: Serverless Framework - AWS Lambda Events - CloudWatch Log
+menuText: Cognito User Pool Trigger
+menuOrder: 10
+description:  Setting up AWS Cognito User Pool Triggers with AWS Lambda via the Serverless Framework
+layout: Doc
+-->
+
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/events/)
```
Comment (hassankhan) [inline docs/providers/aws/events/cognito-user-pool-trigger.md]: Ahh I thought it auto-generated that so I left it alone 😄

File: lib/plugins/aws/lib/naming.test.js [hunk 4]
```diff
@@ -463,4 +470,16 @@ describe('#naming()', () => {
...
+
+  describe('#getLambdaCognitoUserPoolPermissionLogicalId()', () => {
+    it('should normalize the function name and add the standard suffix including the normalized ' +
+      ' pool name and trigger source',
+      () => {
+        expect(sdk.naming.getLambdaCognitoUserPoolPermissionLogicalId(
+          'functionName',
+          'poolId',
+          'TriggerSource'
+        )).to.equal('FunctionNameLambdaPermissionCognitoUserPoolTriggerSourcePoolIdTriggerSource');
```
Comment (hassankhan) [inline lib/plugins/aws/lib/naming.test.js]: Can't seem to find any information regarding length limits, the only thing I can find is [this](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html).

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js (hunk 1 above)
Comment (hassankhan) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: Good idea 👍

**Comment by hassankhan** [comment]
Hi @pmuens, I think I've covered everything you requested. The integration tests were indeed failing, my fault for making assumptions. I thought Travis would have failed the build, but it seems integration tests don't run on PRs. 

Also fixed a bug found during the `multiple-pools-multiple-events-multiple-functions` integration test where the wrong Lambda ARN was being added to the pools.

Please let me know if it's all up to scratch 👍

**Comment by hassankhan** [comment]
I've just found what could be an annoying error - occasionally the Cognito User Pool gets created before the Lambdas do. This causes a problem because the User Pool CF template has a `LambdaConfig` property which maps the trigger sources to a Lambda's ARN. 

I think the best way to deal with it might be to add something like: `DependsOn: [{firstLambdaLogicalId}, {secondLambdaLogicalId}]` to a User Pool's CF template, what do you guys think? @pmuens @HyperBrain

**Comment by HyperBrain** [comment]
@hassankhan @pmuens Regarding the dependency declaration I think this is a good solution to ensure that everything works on deployment.

File: docs/providers/aws/events/cognito-user-pool-trigger.md (hunk 3 above)
Comment (pmuens) [inline docs/providers/aws/events/cognito-user-pool-trigger.md]: 😄 Yes, for some reason it's actually not auto-generated. Don't know why though 😄

**Comment by pmuens** [comment]
@hassankhan just wanted to do a quick follow-up on this PR.

Do you need any help here? Can we do smth. so that it's ready to be merged in v1.15?

**Comment by hassankhan** [comment]
@pmuens Sorry I haven't had time to update it. Weirdly the integration tests seemed to run fine on my machine (I ran them with this: `node_modules/.bin/jest cognito-user-pool-trigger`). I also haven't  added the `DependsOn` attribute either.

I'm not sure how much time I'll have this week as I have work commitments, I'll try my best 👍

**Comment by pmuens** [comment]
Thanks for the update @hassankhan 👍 

I'll debug this integration test issue again this week and will also look into the `DependsOn` property.

> I'm not sure how much time I'll have this week as I have work commitments, I'll try my best 👍

No worries! Don't want to hurry you. Just wanted to make sure that we can get this into the next release since it's a super valuable feature.

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 5]
```diff
@@ -103,12 +103,16 @@ class AwsCompileCognitoUserPoolEvents {
 
       const userPoolLogicalId = this.provider.naming.getCognitoUserPoolLogicalId(poolName);
 
+      const DependsOn = _.map(currentPoolTriggerFunctions, (value) => this
+        .provider.naming.getLambdaLogicalId(value.functionName));
+
       const userPoolTemplate = {
         Type: 'AWS::Cognito::UserPool',
```
Comment (HyperBrain) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: I'm not sure if it is wise to create the user pool for the user. There are lots of other User Pool settings that have to be made available in that case (e.g. password complexity, MFA, any many more). The standard settings are not usable in a production environment.

Comment (pmuens) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: Thanks for the feedback @HyperBrain 👍 

I'm not too familiar with Cognito User Pools. Should we keep it here or remove it? I'm always in favor of having a default (even if it's super rudimentary).

**Comment by hassankhan** [comment]
@pmuens Thank you so much for taking the time to finish off this PR!! 👍

**Comment by pmuens** [comment]
> @pmuens Thank you so much for taking the time to finish off this PR!! 👍

Thanks! Shoutouts to you to get this started 💪

**Comment by keshavkaul** [comment]
@pmuens I have already updated sls to 1.5.1, and I can't seem to successfully test cognitoUserPool event. When deploying, I don't get any errors. And going in cognito console, I don't see the trigger link setup with the lambda.

Also, it might help to clarify what `MyUserPool` exactly is - pool Id or pool name. I have tried both but no success.

**Comment by pmuens** [comment]
@keshavkaul thanks for the comment.

This feature was added to Serverless `v1.15.0` and is not available in Serverless `v1.5.x`

Could you open up a new issue which shows your `serverless.yml` and gists of your generated CloudFormation files if you're on `v1.15.0`?

Thanks in advance!

**Comment by keshavkaul** [comment]
@pmuens Sorry I have v1.15.1

**Comment by jamesladd** [comment]
Is there some example code to go with this?

**Comment by dbligh** [comment]
I am struggling trying to get this to work correctly.

I've got a UserPool defined in my resources, and have the lambda trigger defined under functions. But when I view the user pool in the console, it has no triggers assigned to the user pool. Both are created separately, but unlinked.

Here's a sample project I'm testing this with (latest version of serverless 1.25.0):

`service: testService

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello
    events:
      - cognitoUserPool:
        pool: SampleUserPool
        trigger: CustomMessage
resources:
  Resources:
    CognitoUserPoolSampleUserPool:
      Type: AWS::Cognito::UserPool
`

**Comment by jamesladd** [comment]
You might be experiencing the same issue I had.

Can you also share your resources section where you define the CognitoUserPool as this w help resolve the issue

**Comment by dbligh** [comment]
I think I finally resolved this, 2 days of hair pulling :)

```service: testService

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello
resources:
  Resources:
    CognitoUserPoolSampleUserPool:
      Type: AWS::Cognito::UserPool
      DependsOn: HelloLambdaFunction
      Properties:
        LambdaConfig:
          CustomMessage: 
            Fn::GetAtt: [ HelloLambdaFunction, Arn ]
```

The 'hello' function gets turned into 'HelloLambdaFunction' (case sensitive) in the Cfn script, so that's how you need to reference it. I can now see both the UserPool and the Trigger properly configured.

Hope this helps anyone else that has been struggling on this.

**Comment by robfapadmi** [comment]
@dbligh your solution certainly works, but that was possible before this PR!  I am also unable to get the new event trigger working.  My serverless.yml looks like this:

```
functions:
  customCognitoMessages:
    handler: src/CustomCognitoMessages.handler
    events:
      - cognitoUserPool:
        pool: [my-user-pool-id]
        trigger: CustomMessage


resources:
  Resources:
    cognitoIdentityPool:
      Type: AWS::Cognito::IdentityPool
      Properties:
        AllowUnauthenticatedIdentities: false
        IdentityPoolName:
          [my-identity-pool-id]
```

I'm then checking the Triggers section in the Cognito AWS console, but it remains empty.  Does anyone have any ideas as to why this is not working?

**Comment by hassankhan** [comment]
Man I really messed up when writing the docs ... everyone's very confused 😞 

@robfapadmi:

```yaml
functions:
  customCognitoMessages:
    handler: src/CustomCognitoMessages.handler
    events:
      - cognitoUserPool:
        pool: [my-user-pool-id]
        trigger: CustomMessage


resources:
  Resources:
    CognitoIdentityPool[my-identity-pool-id]:
      Type: AWS::Cognito::IdentityPool
      Properties:
        AllowUnauthenticatedIdentities: false
```

**Comment by jamesladd** [comment]
Hassan, That is much more clear.
The implicitly (magical) prefix is hard to remember / see in the doco.

Thank you for taking the time for this update

**Comment by hassankhan** [comment]
If anyone would be so kind as to make a PR that explains this perhaps in a clearer way?

@jamesladd My pleasure

**Comment by dbligh** [comment]
It turns out that the cause of most of my headaches was because of a plugin I was using to split stacks (serverless-nested-stack). Was causing a bunch of dependency issues for me that I couldn't work around.

My resolution was to isolate all my cognito components into a separate project, and create stack outputs, and refer to them in my main project. The other benefit to this being that it keeps my resource count down in my main project.

**Comment by janoist1** [comment]
@hassankhan Unfortunately your example config doesn't work for me, it won't set the `PostConfirmation` trigger.

**Comment by JaderCM** [comment]
Same here. Following [this document](https://serverless.com/framework/docs/providers/aws/events/cognito-user-pool/) the lambda function and cognito pool are created but not linked (the combo to this trigger is empty).

```yaml
service: "CognitoTest"
provider:
  name: aws
  runtime: nodejs6.10

functions:
  CognitoTriggerPostConfirmation:
    handler: Cognito/Triggers.PostConfirmation
    events:
      - cognitoUserPool:
        pool: TestUserPool
        trigger: PostConfirmation

resources:
  Resources:
    CognitoUserPoolTestUserPool:
      Type: AWS::Cognito::UserPool
```

serverless 1.32.0

**Comment by jamesladd** [comment]
I managed to get this working, w post my yml shortly

**Comment by jamesladd** [comment]
Here you go. I hope this helps

Note the naming of the User Pool: CognitoUserPoolCognitoUserPool

resources:
  Resources:
    CognitoUserPoolCognitoUserPool:
      Type: AWS::Cognito::UserPool
      Properties:
        UserPoolName: ${self:provider.environment.USER_POOL_NAME}
        MfaConfiguration: "OFF"
        AutoVerifiedAttributes:
        - email
        AliasAttributes:
        - email
        EmailVerificationMessage: "Your Account confirmation code is {####}"
        EmailVerificationSubject: "Your Account confirmation code"
        EmailConfiguration:
          ReplyToEmailAddress: "xxx@yyy.com"
          SourceArn: "arn:aws:ses:us-east-1:000:identity/xxx@yyy.com"
        Policies:
          PasswordPolicy:
            MinimumLength: 8
            RequireLowercase: true
            RequireNumbers: true
            RequireSymbols: false
            RequireUppercase: true
        Schema:
        - Name: nickname
          AttributeDataType: String
          Mutable: true
          Required: true
        - Name: name
          AttributeDataType: String
          Mutable: true
          Required: true
        - Name: email
          AttributeDataType: String
          Mutable: true
          Required: true
        - Name: phone_number
          AttributeDataType: String
          Mutable: true
          Required: true
        - Name: accountNumber
          AttributeDataType: String
          Mutable: true

    CognitoUserPoolClient:
      Type: AWS::Cognito::UserPoolClient
      Properties:
        ClientName: ${self:provider.environment.USER_POOL_CLIENT_NAME}
        GenerateSecret: false
        UserPoolId:
          Ref: CognitoUserPoolCognitoUserPool

    CognitoIdentityPool:
      Type: AWS::Cognito::IdentityPool
      Properties:
        IdentityPoolName: ${self:provider.environment.IDENTITY_POOL_NAME}
        AllowUnauthenticatedIdentities: true
        CognitoIdentityProviders:
        - ClientId:
            Ref: CognitoUserPoolClient
          ProviderName:
            Fn::GetAtt: [CognitoUserPoolCognitoUserPool, ProviderName]

    CognitoIdentityPoolRoles:
      Type: AWS::Cognito::IdentityPoolRoleAttachment
      Properties:
        IdentityPoolId:
          Ref: CognitoIdentityPool
        Roles:
          authenticated:
            Fn::GetAtt: [CognitoAuthRole, Arn]
          unauthenticated:
            Fn::GetAtt: [CognitoUnauthRole, Arn]

    CognitoAuthRole:
      Type: AWS::IAM::Role
      Properties:
        RoleName: ${self:provider.environment.APP_AUTH_ROLE_NAME}
        Path: /
        AssumeRolePolicyDocument:
          Version: "2012-10-17"
          Statement:
          - Effect: "Allow"
            Principal:
              Federated: "cognito-identity.amazonaws.com"
            Action:
            - sts:AssumeRoleWithWebIdentity
            - sts:AssumeRole
            Condition:
              StringEquals:
                "cognito-identity.amazonaws.com:aud":
                  Ref: CognitoIdentityPool
              "ForAnyValue:StringLike":
                "cognito-identity.amazonaws.com:amr": authenticated
        Policies:
        - PolicyName: "CognitoAuthorizedPolicy"
          PolicyDocument:
            Version: "2012-10-17"
            Statement:
            - Effect: "Allow"
              Action:
              - mobileanalytics:PutEvents
              - cognito-sync:*
              - cognito-identity:*
              Resource: "*"
            - Effect: "Allow"
              Action:
              - lambda:InvokeFunction
              - lambda:InvokeAsync
              Resource: "*"
            - Effect: Allow
              Action:
              - iot:*
              Resource: "*"

    CognitoUnauthRole:
      Type: AWS::IAM::Role
      Properties:
        RoleName: ${self:provider.environment.APP_UNAUTH_ROLE_NAME}
        Path: /
        AssumeRolePolicyDocument:
          Version: "2012-10-17"
          Statement:
          - Effect: "Allow"
            Principal:
              Federated: "cognito-identity.amazonaws.com"
            Action:
            - sts:AssumeRoleWithWebIdentity
            - sts:AssumeRole
            Condition:
              StringEquals:
                "cognito-identity.amazonaws.com:aud":
                  Ref: CognitoIdentityPool
              "ForAnyValue:StringLike":
                "cognito-identity.amazonaws.com:amr": unauthenticated
        Policies:
        - PolicyName: "CognitoUnauthorizedPolicy"
          PolicyDocument:
            Version: "2012-10-17"
            Statement:
            - Effect: "Allow"
              Action:
              - mobileanalytics:PutEvents
              - cognito-sync:*
              - cognito-identity:*
              Resource: "*"
            - Effect: "Allow"
              Action:
              - lambda:InvokeFunction
              - lambda:InvokeAsync
              Resource: "*"
            - Effect: Allow
              Action:
              - iot:*
              Resource: "*"

**Comment by JaderCM** [comment]
Thank you for reply @jamesladd, but your configuration not have a trigger of Cognito Pool.

I see some messages customized like "EmailVerificationMessage" or "EmailVerificationSubject", polices to invoke lambda functions.

My and @janoist1 case is a trigger when the user confirm the registration process. The combo that we are talking about is this:

screenshot_1

**Comment by jamesladd** [comment]
Oops, here you go.
This when run sets up my Cognito Pool AND registers the functions to be
tiggered.

functions:

  hello:
    handler: src/echo.hello
    events:
      - http: GET hello

  userPreSignUp:
    description: User pre-signup trigger Handler.
    handler: src/api.regoPreSignUp
    memorySize: ${self:provider.environment.MEMORY_DEFAULT}
    timeout: ${self:provider.environment.TIMEOUT_DEFAULT}
    private: true
    events:
      - cognitoUserPool:
          pool: CognitoUserPool
          trigger: PreSignUp
    role: UserRegistrationRole

  userCustomMessage:
    description: User Custom Message trigger Handler.
    handler: src/api.regoCustomMessage
    memorySize: ${self:provider.environment.MEMORY_DEFAULT}
    timeout: ${self:provider.environment.TIMEOUT_DEFAULT}
    private: true
    events:
      - cognitoUserPool:
          pool: CognitoUserPool
          trigger: CustomMessage
    role: UserRegistrationRole

  userPostConfirmation:
    description: User post-confirmation trigger Handler.
    handler: src/api.regoPostConfirmation
    memorySize: ${self:provider.environment.MEMORY_DEFAULT}
    timeout: ${self:provider.environment.TIMEOUT_DEFAULT}
    private: true
    events:
      - cognitoUserPool:
          pool: CognitoUserPool
          trigger: PostConfirmation
    role: UserRegistrationRole

**Comment by JaderCM** [comment]
What is the configuration of your role "UserRegistrationRole"? I'm trying reproduce here your example.

**Comment by JaderCM** [comment]
Thank you @jamesladd. I compared my yaml file with your and I found the problem with mine.

This yaml file not works:
```yaml
functions:
  CognitoTriggerPostConfirmation:
    handler: Cognito/Triggers.PostConfirmation
    private: true
    events:
      - cognitoUserPool:
        pool: TestUserPool
        trigger: PostConfirmation

resources:
  Resources:
    CognitoUserPoolTestUserPool:
      Type: AWS::Cognito::UserPool
```

But this code works:
```yaml
functions:
  CognitoTriggerPostConfirmation:
    handler: Cognito/Triggers.PostConfirmation
    private: true
    events:
      - cognitoUserPool:
          pool: TestUserPool
          trigger: PostConfirmation

resources:
  Resources:
    CognitoUserPoolTestUserPool:
      Type: AWS::Cognito::UserPool
```

The difference is 2 spaces before the section inside of the property cognitoUserPool. The [documentation](https://serverless.com/framework/docs/providers/aws/events/cognito-user-pool#overriding-a-generated-user-pool) show this and I lost. My mistake.

@janoist1 maybe is the same for you.

**Comment by JaderCM** [comment]
@hassankhan it's possible to reference the cognito user pool with another way than with prefix "CognitoUserPool"?

My current stage have a CognitoUserPool with users and if I change the logical id, another cognito user pool will be created.

So, how can I reference my current CognitoUserPool in this lambda?

My case:

```yaml
functions:
  CognitoTriggerPostConfirmation:
    handler: Cognito/Triggers.PostConfirmation
    private: true
    events:
      - cognitoUserPool:
          pool: [how-to-get-MyExistentUserPool?]
          trigger: PostConfirmation

resources:
  Resources:
    MyExistentUserPool:
      Type: AWS::Cognito::UserPool
```

**Comment by JaderCM** [comment]
@hassankhan I see here the function that you are using to retrieve the Cognito User Pool by the Logical Id (getCognitoUserPoolLogicalId) and I ask you for what reason you did this?

My question is related with the cases of existent Cognito User Pool and we can not change the Logical Id to use this feature to link a Lambda Function to a Cognito User Pool trigger.

In this case, if we use the same values of "User Pool Logical Id" and property "cognitoUserPool.pool" the link will works fine.

**Comment by jamesladd** [comment]
Good to hear

**Comment by hassankhan** [comment]
Hi @JaderCM, sorry for the delayed response. 

> it's possible to reference the cognito user pool with another way than with prefix "CognitoUserPool"?

Unfortunately no, since this is following the same [Serverless naming patterns](https://serverless.com/framework/docs/providers/aws/guide/resources#aws-cloudformation-resource-reference) that all other resources do.

> So, how can I reference my current CognitoUserPool in this lambda?

As I understand it, you cannot reference a pre-existing user pool. You would have to use the one generated by the framework.

>  I see here the function that you are using to retrieve the Cognito User Pool by the Logical Id (getCognitoUserPoolLogicalId) and I ask you for what reason you did this?

Not entirely sure what you're referring to...

For any other questions, I'd strongly suggest opening a new issue instead of commenting on this PR 😄. That way, there'd be more visibility rather than on this PR which only has a few eyes on it 👍

**Comment by JaderCM** [comment]
Hi @hassankhan, thanks for reply.

> As I understand it, you cannot reference a pre-existing user pool. You would have to use the one generated by the framework.

My current (pre-existing) user pool was created by the serverless framework. The problem is that now I need to link this pool with my lambda by trigger and I cannot.

> Not entirely sure what you're referring to...

My question is about if is possible to reference user pool using just the logical id without prefix.

I opened the issue #5401 to discuss about.

**Comment by iamdavidmartin** [comment]
this pull requests has caused me a lot of pain. in hindsight, i wish i would have never touched this.

1. silent failures. this code silently fails when you do not indent `cognitoUserPool` properties like `pool` and `trigger`. i am not the only person who has strugged with this. it ought to tell you that you've misconfigured the parameters.

- https://github.com/serverless/serverless/pull/3657#issuecomment-430207100
- https://github.com/serverless/serverless/issues/6356

2. circular dependencies. this code creates them. maybe that's just a problem with cognito triggers, but it's simple to create triggers without the dependencies.
- https://forum.serverless.com/t/cognito-user-pool-trigger-event/3589
- https://github.com/serverless/serverless/issues/5401#issuecomment-434346604

3. can't reference the UserPool by the actual name yml name of the user pool. It's undocumented, but you have to figure out through extensive googling that if your user pool is called "CognitoUserPoolUsers", the `pool` property should get "Users". 

i was able to implement the correct trigger without any of these headaches with a single line of code elsewhere.

 `PostConfirmation: "arn:aws:lambda:#{AWS::Region}:#{AWS::AccountId}:function:${self:service}-${self:provider.stage}-postConfirmation"`

🤦🏻‍♂️🤦🏻‍♂️🤦🏻‍♂️

**Pull request merged**